File: code/interviewQs/leetcode/removeDupesFromSortedList.py
"""
Given a sorted linked list, delete all nodes that have duplicate numbers, leaving only distinct numbers from the original list.

Return the linked list sorted as well.

Example 1:

Input: 1->2->3->3->4->4->5
Output: 1->2->5
Example 2:

Input: 1->1->1->2->3
Output: 2->3
"""

import pdb
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def buildList(self, nums) -> ListNode:
        head = None
        cur = None
        for n in nums:
            if head == None:
                head = ListNode(n)
                cur = head
            else:
                cur.next = ListNode(n)
                cur = cur.next

        return head

    def printList(self, head: ListNode):
        print('|', end='')
        while head != None:
            print('{}->'.format( head.val), end='');
            head = head.next
        print('|', end='')
        print('')


    def deleteDuplicates(self, head: ListNode) -> ListNode:
        #pdb.set_trace()
        prev, cur = None, None

        while True: #delete dupes in prefix
            shouldDelete = False
            if (head == None) or (head.next == None):
                return head
            prev = head
            cur = head.next

            while (cur != None) and (prev.val == cur.val):
                shouldDelete = True
                prev = cur
                cur = cur.next

            if shouldDelete:
                head = cur
            else:
                break

        #now prev != cur, prev point to final head
        while cur != None:
            shouldDelete = False
            while (cur.next != None) and (cur.next.val == cur.val):
                shouldDelete = True
                cur = cur.next
            
            if shouldDelete:
                prev.next = cur.next #delete sames
                if cur.next != None:
                    if (cur.next.next != None) and (cur.next.val != cur.next.next.val):
                        #invariant is kept prev != cur. 
                        prev = cur.next
                        cur = cur.next.next 
                    else:
                        cur = cur.next
                else:
                    break
            else:
                prev = cur
                cur = cur.next
        
        return head

def test():

    #inputs = [
    #    [1,2,2],
    #]
    inputs = [
        [],
        [1],
        [1,2,2],
        [1,2,3,3,4,4,5],
        [1,1,1,2,3],
    ]

    sol = Solution()
    
    for inp in inputs:
        lst = sol.buildList(inp)
        print('original list')
        sol.printList(lst)
        lst = sol.deleteDuplicates(lst)
        print('remove dupes list')
        sol.printList(lst)


if __name__ == "__main__":
    test()


        

File: code/interviewQs/leetcode/removeDupesFromSortedList.cpp
/**
Given a sorted linked list, delete all nodes that have duplicate numbers, leaving only distinct numbers from the original list.

Return the linked list sorted as well.

Example 1:

Input: 1->2->3->3->4->4->5
Output: 1->2->5
Example 2:

Input: 1->1->1->2->3
Output: 2->3
**/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        /*
        idea, 
        1st step remove leading sames, move header to 1st elem post sames
        2nd step. there r two distinct elements. use 2 ptrs. prev, curr
        for curr run loop to delete if there r sames.
        if not prev,curr = curr, curr->next
        */

        ListNode* prev;
        ListNode* curr;
        
        //process prefix of sames
        // exit cond. a. curr is null (finished list), b. prev != curr
        while (true) { 
            if ( (head == nullptr) || (head->next == nullptr) ) return head;
            prev = head;
            curr = head->next;
            
            if ( (curr != nullptr) && (curr->val  == prev->val ) ) {
                curr = curr->next;
                while ((curr != nullptr) && (curr->val == prev->val)) {
                    curr = curr->next; 
                }
                head = curr;
            }
            else {
                break;
            }
        }

        if (curr == nullptr) return head;
        //maintain invarianta: prev and curr are distinct.
        while (curr->next != nullptr) {
            if (curr->next->val == curr->val) {
                while (( curr->next != nullptr) && (curr->next->val == curr->val)) {
                    curr = curr->next;
                } 
                prev->next = curr;
            }
            else {
                prev = curr;
                curr = curr->next;
            } 
        }

        return head;
    }
};